articles/actions.md

```diff
@@ -0,0 +1,363 @@
+
+### Action Server
+
+An action server provides an action.
+Like topics and services, an action server has a name and a type.
+There is only one server per name.
+
+It is responsible for:
+
+- advertising the action to other ROS entities
+- accepting or rejecting goals from one or more action clients
+- executing the action when a goal is received and accepted
+- optionally providing feedback about the progress of all executing actions
+- optionally handling requests to cancel one or more actions
+- sending the result of an action, including whether it succeed, failed, or was canceled, to the client when the action completes
+
+### Action Client
+
+An action client sends a goal (an action to be performed) and monitors its progress.
+There may be multiple clients per server; however, it is up to the server to decide how goals from multiple clients will be handled.
+
+It is responsible for:
+
+- sending a goal to the action server
+- optionally monitoring the feedback for a goal from the action server
+- optionally monitoring the status for a goal from the action server
+- optionally requesting that the action server cancel an active goal
+- optionally checking the result for a goal received from the action server
+
+## Differences between ROS 1 and ROS 2 actions
+
+### First Class Support
+
+In ROS 1, actions are implemented as a separate library, [actionlib](http://wiki.ros.org/actionlib) that is built on top of the client libraries.
+This was done to avoid increasing the work required to create a client library in a new language, but actions turned out to be very important to packages like the [Navigation Stack](http://wiki.ros.org/navigation) and [MoveIt!](https://moveit.ros.org/)<sup>[1](#separatelib)</sup>.
+
+In ROS 2, actions will be included in the client library implementations.
+The work of writing a client library in a new language will be reduced by creating a common implementation in C.
+
+### Services used for Actions
+
+In ROS 1, actions were implemented using a set of topics under a namespace taken from the action name.
+ROS 1 services were not used because they are inherently synchronous and actions need to be asynchronous.
+Actions also needed to send status/feedback and be cancelable.
+
+In ROS 2, services are asynchronous in the common C implementation, so actions will use a combination of services and topics.
+
+### Goal Identifiers
+
+In ROS 1, an action client can create a goal ID when submitting a goal.
+This potentially leads to scenarios where mutliple clients independently generate the same goal ID.
+If an empty goal ID is sent, the action server will create one instead, which is not very useful since the client has no way to know the goal ID.
+
+In ROS 2, action clients will be the sole entities responsible for generating the goal ID.
+This way clients always know the goal ID for an action.
+Futhermore, a UUID will be used for each goal to mitigate the issue of goal ID collision across multiple clients.
```

Comment:
This was the original proposal because of the nice ability to guarantee goal ID uniqueness for each goal request, but the proposal was updated because it seemed easier to have the client generate a UUID and allow for relaxation of the constraint that an action client needs to get a goal response.

AFAIU, if the application developer decides to use unreliable QoS settings for the goal service, then we can't guarantee a goal service response is received by the action client. But, the action server may have very well accepted and started executing a goal. In this scenario, the client can still interact with the action goal as it can look for the UUID it generated in the feedback/status topics and make result/cancel requests.

If we have the server generate the UUID, then in the scenario I described, the action client does not know the UUID and therefore cannot know the state of the goal or make cancel/result requests.

In summary, having the client generate a UUID allows for more robust usage of actions when working with unreliable communications. I can add this rationale to the design doc if it makes sense.

Reply:
I don't agree with the rationale. I think that, at the very least, reliable QoS should be used for the goal creation service, even if everything else is best-effort. Do you have a use case where a fire-and-forget action needs to be created where even the response to creating the action cannot be reliable?

Reply:
> I think that, at the very least, reliable QoS should be used for the goal creation service, even if everything else is best-effort. Reliable QoS can handle an unreliable communications layer.

Agreed, the default QoS settings should be reliable. The plan is to allow a user to change the QoS settings.

> In fact, I just noticed that the "send goal" service states:
> >The QoS settings of this service should be set so the client is guaranteed to receive a response or an action could be executed without a client being aware of it.

That warning was added when the proposal had the server creating the goal ID. If the user changes the QoS to be best-effort on the "send goal" service then there is no way for a client to check if a goal is accepted. The  change to the client creating the goal ID would allow removing that warning.

Reply:
I was thinking that we'd expose QoS settings for all of the services/topics that comprise an action so that users have full control. Certainly the default for the send goal service should guarantee a receive and response (which I think is the default for ROS 2 services).

I don't have a specific use-case in mind, but if we allow users to configure the send goal service to be best effort, then I think having the client generate the UUID makes sense. Alternatively, we can choose to restrict configuring the send goal QoS settings and then I'm okay with generating the UUID at either end, server or client. The ID uniqueness can still be maintained by the action server by not accepting a goal request for an ID that has already been accepted.

Reply:
Here is the original motivation for having the action  client generate the goal ID: https://github.com/ros2/design/pull/193#discussion_r225267546

Basically, being able to fire-and-forget and be able to correlate goals with other libraries/systems sounds reasonable.

I believe the implementation of the action server doesn't change too much in either case. Whether it's the server or client generating the goal ID, the action server is still responsible for handling a potential race where concurrent goal requests arrive and it needs to ensure that each goal ID is unique.

I didn't want to leave this thread without a conclusion before merging. I can add to the alternatives section of the doc so the motivation behind the decision isn't lost.

Reply:
OK, I can accept that motivation.

Reply:
Added motivation to doc b7ac32a